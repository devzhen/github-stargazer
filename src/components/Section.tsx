import * as React from 'react';
import styled from 'styled-components/native';

interface SectionProps {
  offsetTop?: number;
  offsetBottom?: number;
  offsetLeft?: number;
  offsetRight?: number;
  clientLeft?: number;
  clientRight?: number;
  clientTop?: number;
  clientBottom?: number;
  fluid?: boolean;
  stretchContent?: boolean;
  children?: React.ReactNode;
  centerContentHorizontally?: boolean;
  centerContentVertically?: boolean;
  flexEndContentVertically?: boolean;
  flexStartContentHorizontally?: boolean;
  bgColor: string;
  width?: string | number;
  height?: string | number;
  borderRadius?: number;
  borderColor?: string | null;
  borderTopWidth?: number;
  borderLeftWidth?: number;
  borderRightWidth?: number;
  borderBottomWidth?: number;
  customStyle?: any;
  directionRow?: boolean;
}

const Section = ({
  offsetTop,
  offsetBottom,
  offsetLeft,
  offsetRight,
  clientLeft,
  clientRight,
  clientTop,
  clientBottom,
  children,
  fluid,
  stretchContent,
  centerContentHorizontally,
  centerContentVertically,
  flexEndContentVertically,
  flexStartContentHorizontally,
  bgColor,
  width,
  height,
  borderRadius,
  borderColor,
  borderTopWidth,
  borderLeftWidth,
  borderRightWidth,
  borderBottomWidth,
  customStyle,
  directionRow,
}: SectionProps) => (
  <Container
    directionRow={directionRow}
    bgColor={bgColor}
    offsetTop={offsetTop}
    offsetBottom={offsetBottom}
    offsetLeft={offsetLeft}
    offsetRight={offsetRight}
    clientLeft={clientLeft}
    clientRight={clientRight}
    clientTop={clientTop}
    clientBottom={clientBottom}
    fluid={fluid}
    stretchContent={stretchContent}
    centerContentHorizontally={centerContentHorizontally}
    centerContentVertically={centerContentVertically}
    flexEndContentVertically={flexEndContentVertically}
    flexStartContentHorizontally={flexStartContentHorizontally}
    width={width}
    height={height}
    borderRadius={borderRadius}
    borderColor={borderColor}
    borderTopWidth={borderTopWidth}
    borderLeftWidth={borderLeftWidth}
    borderRightWidth={borderRightWidth}
    borderBottomWidth={borderBottomWidth}
    style={customStyle}
  >
    {children}
  </Container>
);

Section.defaultProps = {
  offsetTop: 0,
  offsetBottom: 0,
  offsetLeft: 0,
  offsetRight: 0,
  clientLeft: 0,
  clientRight: 0,
  clientTop: 0,
  clientBottom: 0,
  fluid: false,
  stretchContent: false,
  centerContentHorizontally: false,
  centerContentVertically: false,
  flexStartContentHorizontally: false,
  bgColor: 'transparent',
  borderRadius: 0,
  borderColor: null,
  customStyle: {},
};

const Container = styled.View<SectionProps>`
  flex-direction: ${(props) => (props.directionRow ? 'row' : 'column')};
  margin-top: ${(props) => props.offsetTop};
  margin-bottom: ${(props) => props.offsetBottom};
  margin-left: ${(props) => props.offsetLeft};
  margin-right: ${(props) => props.offsetRight};
  padding-left: ${(props) => props.clientLeft};
  padding-right: ${(props) => props.clientRight};
  padding-top: ${(props) => props.clientTop};
  padding-bottom: ${(props) => props.clientBottom};
  ${(props) => props.fluid && 'flex: 1'};
  ${(props) => props.stretchContent && 'justify-content: space-between;'};
  ${(props) => props.centerContentHorizontally && 'align-items: center;'};
  ${(props) => props.centerContentVertically && 'justify-content: center;'};
  ${(props) => props.flexEndContentVertically && 'justify-content: flex-end;'};
  ${(props) => props.width && `width: ${props.width}`};
  ${(props) => props.height && `height: ${props.height}`};
  ${(props) => props.borderRadius && `border-radius: ${props.borderRadius}`};
  ${(props) => props.borderColor && `border-color: ${props.borderColor}`};
  ${(props) => props.borderColor && `border-width: 1`};
  ${(props) =>
    props.borderTopWidth && `border-top-width: ${props.borderTopWidth}`};
  ${(props) =>
    props.borderLeftWidth && `border-left-width: ${props.borderLeftWidth}`};
  ${(props) =>
    props.borderRightWidth && `border-right-width: ${props.borderRightWidth}`};
  ${(props) =>
    props.borderBottomWidth &&
    `border-bottom-width: ${props.borderBottomWidth}`};
  background-color: ${(props) => props.bgColor};
  ${(props) =>
    props.flexStartContentHorizontally &&
    props.directionRow &&
    `justify-content: flex-start`};
  ${(props) =>
    props.flexStartContentHorizontally &&
    !props.directionRow &&
    `align-items: flex-start`};
`;

export default Section;
