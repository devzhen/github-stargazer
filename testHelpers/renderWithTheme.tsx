import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';
import 'jest-styled-components';

import theme from '@theme/index';

const renderWithTheme = (component: any) =>
  renderer.create(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

export default renderWithTheme;
