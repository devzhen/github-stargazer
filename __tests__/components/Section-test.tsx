import React from 'react';
import { View, Text } from 'react-native';
import 'jest-styled-components';

import Section from '@components/Section';

// @ts-ignores
import renderWithTheme from '@testHelpers/renderWithTheme';

describe('Section common component', () => {
  it('renders correctly, 1 test', () => {
    const tree = renderWithTheme(
      <Section fluid clientLeft={8} clientRight={8}>
        <View>
          <Text>TEST</Text>
        </View>
      </Section>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly, 2 test', () => {
    const tree = renderWithTheme(
      <Section centerContentVertically>
        <View>
          <Text>TEST</Text>
        </View>
      </Section>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly, 3 test', () => {
    const tree = renderWithTheme(
      <Section fluid borderRadius={22}>
        <View>
          <Text>TEST</Text>
        </View>
      </Section>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
