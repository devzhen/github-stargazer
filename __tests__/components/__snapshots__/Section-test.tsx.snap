// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section common component renders correctly, 1 test 1`] = `
<View
  bgColor="transparent"
  borderColor={null}
  borderRadius={0}
  centerContentHorizontally={false}
  centerContentVertically={false}
  clientBottom={0}
  clientLeft={8}
  clientRight={8}
  clientTop={0}
  flexStartContentHorizontally={false}
  fluid={true}
  offsetBottom={0}
  offsetLeft={0}
  offsetRight={0}
  offsetTop={0}
  stretchContent={false}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "paddingBottom": 0,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 0,
      },
      Object {},
    ]
  }
>
  <View>
    <Text>
      TEST
    </Text>
  </View>
</View>
`;

exports[`Section common component renders correctly, 2 test 1`] = `
<View
  bgColor="transparent"
  borderColor={null}
  borderRadius={0}
  centerContentHorizontally={false}
  centerContentVertically={true}
  clientBottom={0}
  clientLeft={0}
  clientRight={0}
  clientTop={0}
  flexStartContentHorizontally={false}
  fluid={false}
  offsetBottom={0}
  offsetLeft={0}
  offsetRight={0}
  offsetTop={0}
  stretchContent={false}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "flexDirection": "column",
        "justifyContent": "center",
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "paddingBottom": 0,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 0,
      },
      Object {},
    ]
  }
>
  <View>
    <Text>
      TEST
    </Text>
  </View>
</View>
`;

exports[`Section common component renders correctly, 3 test 1`] = `
<View
  bgColor="transparent"
  borderColor={null}
  borderRadius={22}
  centerContentHorizontally={false}
  centerContentVertically={false}
  clientBottom={0}
  clientLeft={0}
  clientRight={0}
  clientTop={0}
  flexStartContentHorizontally={false}
  fluid={true}
  offsetBottom={0}
  offsetLeft={0}
  offsetRight={0}
  offsetTop={0}
  stretchContent={false}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 22,
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "paddingBottom": 0,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 0,
      },
      Object {},
    ]
  }
>
  <View>
    <Text>
      TEST
    </Text>
  </View>
</View>
`;
